src/app/models/workspace.server.model.ts

```diff
@@ -0,0 +1,90 @@
+import { Mongoose, Schema } from 'mongoose'
+
+import { IWorkspaceModel, IWorkspaceSchema } from '../../types'
+
+import getFormModel from './form.server.model'
+import getUserModel from './user.server.model'
+
+export const WORKSPACE_SCHEMA_ID = 'Workspace'
+
+const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
+  const Form = getFormModel(db)
+  const User = getUserModel(db)
+
+  const WorkspaceSchema = new Schema<IWorkspaceSchema, IWorkspaceModel>(
+    {
+      title: {
+        type: String,
+        validate: [
+          /^[a-zA-Z0-9_\-./() &`;'"]*$/,
+          'Workspace title cannot contain special characters',
+        ],
+        required: 'Workspace title cannot be blank',
+        minlength: [4, 'Workspace title must be at least 4 characters'],
+        maxlength: [
+          200,
+          'Workspace title can have a maximum of 200 characters',
+        ],
+        trim: true,
+      },
+      admin: {
+        type: Schema.Types.ObjectId,
+        ref: 'User',
+        required: 'Workspace must have an Admin',
+      },
+      formIds: {
+        type: [Schema.Types.ObjectId],
+        validate: {
+          async validator(this: IWorkspaceSchema) {
+            const user = await User.findById(this.admin)
+            if (!user) return false
+
+            for (const formId of this.formIds) {
+              const form = await Form.findById(formId)
+              if (!form) return false
+
+              const isUserFormAdmin = form.admin.equals(this.admin)
+              const isUserFormCollaborator =
+                form?.permissionList?.find(
+                  (permission) => permission.email === user.email,
+                ) != undefined
+
+              const doesUserHaveFormAccess =
+                isUserFormAdmin || isUserFormCollaborator
+              const isFormIdUnique = this.formIds.includes(formId)
+              const isFormIdValid = doesUserHaveFormAccess && isFormIdUnique
+
+              if (!isFormIdValid) return false
+            }
+
+            return true
+          },
+        },
```

Comment:
i think we should properly distinguish what should be a service-level validation and what should be a model level validation.

model-level validation should be validation that does not rely on business logic, such as formId uniqueness.
service-level validation is probably everything else.

in this validation block, i'd think that checking of permissions should be a service-level validation, as putting that validation here could result in some edge case like when users stop having read permissions  to the form and the db would prevent deletion of that form from the workspace.

Reply:
Right that makes sense. I've moved the check for permissions out of the schema, and only perform `formId` uniqueness check instead.